File: src/Block.ts
import { Hasher } from './Hasher';

export class Block {
  constructor(
    public index: number,
    public timestamp: number,
    public previousHash: string,
    public proof: number
  ) {}

  get hash(): string {
    const properties = {
      index: this.index,
      timestamp: this.timestamp,
      previousHash: this.previousHash,
      proof: this.proof,
    };
    const propertiesString = JSON.stringify(properties);
    return Hasher.sha256(propertiesString);
  }
}

File: create-a-blockchain-go/internal/block/block.go
package block

import (
	"create-a-blockchain-go/internal/hasher"
	"encoding/json"
	"time"
)

type Block struct {
	Index        int       `json:"index"`
	Timestamp    time.Time `json:"timestamp"`
	PreviousHash string    `json:"previousHash"`
	Proof        int       `json:"proof"`
}

func (b *Block) Hash() string {
	properties := map[string]interface{}{
		"index":        b.Index,
		"timestamp":    b.Timestamp,
		"previousHash": b.PreviousHash,
		"proof":        b.Proof,
	}
	propertiesBytes, _ := json.Marshal(properties)
	return hasher.SHA256(string(propertiesBytes))
}

func NewBlock(index int, timestamp time.Time, previousHash string, proof int) *Block {
	return &Block{
		Index:        index,
		Timestamp:    timestamp,
		PreviousHash: previousHash,
		Proof:        proof,
	}
}
